File: examples/java/ConfluenceExample.java
import ai.pipestream.confluence.v1.BodyFormat;
import ai.pipestream.confluence.v1.ConfluenceServiceGrpc;
import ai.pipestream.confluence.v1.GetPageRequest;
import ai.pipestream.confluence.v1.GetPageResponse;
import ai.pipestream.confluence.v1.ListPagesRequest;
import ai.pipestream.confluence.v1.ListPagesResponse;
import ai.pipestream.confluence.v1.ListSpacesRequest;
import ai.pipestream.confluence.v1.Page;
import ai.pipestream.confluence.v1.ProbeConnectionRequest;
import ai.pipestream.confluence.v1.ProbeConnectionResponse;
import ai.pipestream.confluence.v1.Space;
import ai.pipestream.confluence.v1.SyncRequest;
import ai.pipestream.confluence.v1.SyncResponse;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.Iterator;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Reads a Confluence site through the gRPC proxy: probes the connection,
 * lists spaces, streams one space's pages, fetches a page body in storage
 * format, and runs a bounded {@code Sync}, printing the resume cursor a
 * consumer would persist for the next incremental pass.
 *
 * <p>See README.md for build and run commands; the classpath is the
 * repo's own {@code grpc-confluence-service} distribution jars.</p>
 */
public final class ConfluenceExample {

    private ConfluenceExample() {
    }

    public static void main(String[] args) throws InterruptedException {
        String target = System.getenv().getOrDefault("CONFLUENCE_GRPC_TARGET", "localhost:9095");
        ManagedChannel channel = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
        try {
            ConfluenceServiceGrpc.ConfluenceServiceBlockingStub stub =
                    ConfluenceServiceGrpc.newBlockingStub(channel);

            ProbeConnectionResponse probe = stub.probeConnection(
                    ProbeConnectionRequest.newBuilder().setLimit(3).build());
            if (!probe.getOk()) {
                throw new IllegalStateException("probe failed: " + probe.getErrorMessage());
            }
            System.out.println("connected; sample spaces: " + probe.getSpaceKeysList());

            List<Space> spaces = stub.listSpaces(ListSpacesRequest.getDefaultInstance())
                    .getSpacesList();
            for (Space space : spaces) {
                System.out.printf("space %s id=%s name=%s%n",
                        space.getKey(), space.getId(), space.getName());
            }
            if (spaces.isEmpty()) {
                return;
            }
            Space space = spaces.getFirst();

            // Stream every page of the first space; keep the first page id.
            String firstPageId = "";
            Iterator<ListPagesResponse> pages = stub.listPages(
                    ListPagesRequest.newBuilder().setSpaceId(space.getId()).build());
            while (pages.hasNext()) {
                Page page = pages.next().getPage();
                if (firstPageId.isEmpty()) {
                    firstPageId = page.getId();
                }
                System.out.printf("page %s %s%n", page.getId(), page.getTitle());
            }

            if (!firstPageId.isEmpty()) {
                GetPageResponse got = stub.getPage(GetPageRequest.newBuilder()
                        .setId(firstPageId)
                        .setBodyFormat(BodyFormat.BODY_FORMAT_STORAGE_XHTML)
                        .build());
                System.out.printf("page %s storage body: %d chars%n",
                        got.getPage().getTitle(),
                        got.getPage().getBody().getStorage().getValue().length());
            }

            // One bounded sync pass. Persist the resume cursor and pass it
            // back as since_cursor to receive only what changed since.
            int changes = 0;
            String cursor = "";
            Iterator<SyncResponse> sync = stub.sync(SyncRequest.newBuilder()
                    .addSpaceKeys(space.getKey()).build());
            while (sync.hasNext()) {
                SyncResponse event = sync.next();
                switch (event.getEventCase()) {
                    case CHANGE -> changes++;
                    case RESUME_CURSOR -> cursor = event.getResumeCursor();
                    default -> {
                    }
                }
            }
            System.out.printf("sync: %d changes, resume_cursor=%s%n", changes, cursor);
        } finally {
            channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
        }
    }
}

File: examples/python/confluence_example.py
"""Read a Confluence site through the gRPC proxy.

Lists spaces, fetches one page in storage format, streams the space's
pages, and runs a bounded Sync - printing the resume cursor a consumer
would persist for the next incremental pass.

Run ./generate.sh once, then:

    CONFLUENCE_GRPC_TARGET=localhost:9095 python3 confluence_example.py
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "gen"))

import grpc  # noqa: E402

from ai.pipestream.confluence.v1 import (  # noqa: E402
    confluence_service_pb2 as api,
    confluence_service_pb2_grpc as rpc,
    common_pb2 as common,
)


def main() -> None:
    target = os.environ.get("CONFLUENCE_GRPC_TARGET", "localhost:9095")
    with grpc.insecure_channel(target) as channel:
        stub = rpc.ConfluenceServiceStub(channel)

        probe = stub.ProbeConnection(api.ProbeConnectionRequest(limit=3))
        if not probe.ok:
            raise SystemExit(f"probe failed: {probe.error_message}")
        print(f"connected; sample spaces: {list(probe.space_keys)}")

        spaces = stub.ListSpaces(api.ListSpacesRequest()).spaces
        for space in spaces:
            print(f"space {space.key!r} id={space.id} name={space.name!r}")
        if not spaces:
            return

        # Stream every page of the first space; keep the first page id.
        space = spaces[0]
        first_page_id = None
        for response in stub.ListPages(api.ListPagesRequest(space_id=space.id)):
            page = response.page
            first_page_id = first_page_id or page.id
            print(f"page {page.id} {page.title!r}")

        if first_page_id:
            got = stub.GetPage(api.GetPageRequest(
                id=first_page_id,
                body_format=common.BODY_FORMAT_STORAGE_XHTML))
            body = got.page.body.storage.value
            print(f"page {got.page.title!r} storage body: {len(body)} chars")

        # One bounded sync pass. Persist resume_cursor and pass it back as
        # since_cursor to receive only what changed since this run.
        changes = 0
        cursor = ""
        for event in stub.Sync(api.SyncRequest(space_keys=[space.key])):
            kind = event.WhichOneof("event")
            if kind == "change":
                changes += 1
            elif kind == "resume_cursor":
                cursor = event.resume_cursor
        print(f"sync: {changes} changes, resume_cursor={cursor!r}")


if __name__ == "__main__":
    main()
